File: algorithms/2_medium/M/04050_minimum_days_to_score_exactly_n_points/dp.c
#define RESET_COST 1

int minDays(int n) {
    assert(n >= 1 && n <= 100000);
    int width = 1;
    int width_cost = 1;
    int *restrict dp = malloc((size_t)(n + 1) * sizeof(*dp));
    if (dp == NULL) { return -1; }
    dp[0] = 0;
    for (int i = 1; i <= n; i += 1) { dp[i] = INT_MAX; }

    while (width <= n) {
        dp[width] = width_cost;

        for (int i = width + 1; i <= n; i += 1) {
            dp[i] = MIN(dp[i], dp[i - width] + width_cost + RESET_COST);
        }

        width_cost += 1;
        width += width_cost;
    }

    const int res = dp[n];
    free(dp);
    return res;
}

File: algorithms/2_medium/M/04050_minimum_days_to_score_exactly_n_points/dp.rs
impl Solution {
    pub fn min_days(n: i32) -> i32 {
        assert!(n >= 1 && n <= 100_000);
        const RESET_COST: i32 = 1;
        let mut width: i32 = 1;
        let mut width_cost: i32 = 1;
        let mut dp = vec![i32::MAX; (n + 1) as usize];
        dp[0] = 0;

        while width <= n {
            dp[width as usize] = width_cost;

            for i in (width + 1)..=n {
                dp[i as usize] = dp[i as usize].min(dp[(i - width) as usize] + width_cost + RESET_COST);
            }

            width_cost += 1;
            width += width_cost;
        }

        dp[n as usize]
    }
}
